### PR TITLE
DW-4008 Added Terraform Lambda User Expiry

### PR DESCRIPTION
**Pull request by fernando-schmidt-dmw**

### CODE REVIEW

File: terraform/modules/pre-auth-lambda/dynamodb.tf [hunk 1]
```diff
@@ -0,0 +1,10 @@
+resource "aws_dynamodb_table" "dynamodb_table_user" {
+  name         = "user"
```
Comment (dan-hill2802) [inline terraform/modules/pre-auth-lambda/dynamodb.tf:2]: this is quite a vague name, can it be more specific e.g. `analytical-frontend-service-user`? There's a chance we will have other user tables in the future

File: terraform/modules/pre-auth-lambda/iam.tf [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+}
+
+data aws_iam_policy_document policy_logs_user_expiry {
+  statement {
+    actions = [
+        "logs:CreateLogGroup",
+        "logs:CreateLogStream",
+        "logs:PutLogEvents"
+    ]
+    resources = ["arn:aws:logs:*:*:*"]
```
Comment (dan-hill2802) [inline terraform/modules/pre-auth-lambda/iam.tf:49]: it should only have permission to write to it's own log group

File: terraform/modules/pre-auth-lambda/lambda.tf [hunk 3]
```diff
@@ -0,0 +1,13 @@
+resource "aws_lambda_function" "lambda_user_expiry" {
+  filename      = "${path.module}/lambda-user-expiry.zip"
```
Comment (dan-hill2802) [inline terraform/modules/pre-auth-lambda/lambda.tf:2]: we typically have the zip filename match the function name for clarity

File: terraform/modules/pre-auth-lambda/data.tf [hunk 4]
```diff
@@ -0,0 +1,5 @@
+data "archive_file" "lambda_user_expiry_zip" {
+  type        = "zip"
+  source_dir  = "${path.module}/lambda"
+  output_path = "${path.module}/lambda-user-expiry.zip"
```
Comment (dan-hill2802) [inline terraform/modules/pre-auth-lambda/data.tf:4]: might want to rename this to match the function name

File: terraform/modules/pre-auth-lambda/lambda/lambda_function.py [hunk 5]
```diff
@@ -0,0 +1,76 @@
+import os
+import datetime
+import dateutil
+from dateutil.relativedelta import relativedelta
+import boto3
+from boto3.dynamodb.conditions import Key
+
+tableName = os.environ.get("TABLE_NAME")
+regionName = os.environ.get("AWS_REGION")
```
Comment (dan-hill2802) [inline terraform/modules/pre-auth-lambda/lambda/lambda_function.py:9]: should handle the exception if these aren't set

File: terraform/modules/pre-auth-lambda/lambda/lambda_function.py [hunk 6]
```diff
@@ -0,0 +1,76 @@
+import os
+import datetime
+import dateutil
```
Comment (dan-hill2802) [inline terraform/modules/pre-auth-lambda/lambda/lambda_function.py:3]: this isn't used

File: terraform/modules/pre-auth-lambda/lambda/lambda_function.py [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+from dateutil.relativedelta import relativedelta
+import boto3
+from boto3.dynamodb.conditions import Key
+
+tableName = os.environ.get("TABLE_NAME")
+regionName = os.environ.get("AWS_REGION")
+
+dynamodb = boto3.resource("dynamodb", region_name=regionName)
```
Comment (dan-hill2802) [inline terraform/modules/pre-auth-lambda/lambda/lambda_function.py:11]: all variables need to be lower case separated by underscores

File: terraform/modules/pre-auth-lambda/lambda/lambda_function.py [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+table = dynamodb.Table(tableName)
+
+def queryUser(username):
```
Comment (dan-hill2802) [inline terraform/modules/pre-auth-lambda/lambda/lambda_function.py:14]: 2 blank lines required before each function

File: terraform/modules/pre-auth-lambda/variables.tf [hunk 9]
```diff
@@ -0,0 +1,4 @@
+variable "name_prefix" {
+  type        = string
+  description = "(Required) Name prefix for resources created"
+}
```
Comment (dan-hill2802) [inline terraform/modules/pre-auth-lambda/variables.tf:4]: new line at end of file please

File: terraform/modules/pre-auth-lambda/dynamodb.tf (hunk 1 above)
Comment (fernando-schmidt-dmw) [reply terraform/modules/pre-auth-lambda/dynamodb.tf:2]: This is done.

File: terraform/modules/pre-auth-lambda/lambda.tf (hunk 3 above)
Comment (fernando-schmidt-dmw) [reply terraform/modules/pre-auth-lambda/lambda.tf:2]: This is done.

File: terraform/modules/pre-auth-lambda/data.tf (hunk 4 above)
Comment (fernando-schmidt-dmw) [reply terraform/modules/pre-auth-lambda/data.tf:4]: This is done.

File: terraform/modules/pre-auth-lambda/lambda/lambda_function.py (hunk 6 above)
Comment (fernando-schmidt-dmw) [reply terraform/modules/pre-auth-lambda/lambda/lambda_function.py:3]: This is done.

File: terraform/modules/pre-auth-lambda/lambda/lambda_function.py (hunk 7 above)
Comment (fernando-schmidt-dmw) [reply terraform/modules/pre-auth-lambda/lambda/lambda_function.py:11]: This is done.

File: terraform/modules/pre-auth-lambda/lambda/lambda_function.py (hunk 8 above)
Comment (fernando-schmidt-dmw) [reply terraform/modules/pre-auth-lambda/lambda/lambda_function.py:14]: This is done.

File: terraform/modules/pre-auth-lambda/variables.tf (hunk 9 above)
Comment (fernando-schmidt-dmw) [reply terraform/modules/pre-auth-lambda/variables.tf:4]: This is done.

File: terraform/modules/pre-auth-lambda/lambda/lambda_function.py (hunk 5 above)
Comment (fernando-schmidt-dmw) [reply terraform/modules/pre-auth-lambda/lambda/lambda_function.py:9]: This is done, can you please review again after my commit to make sure the solution is good?

**Comment by fernando-schmidt-dmw** [comment]
> you should run [black](https://github.com/psf/black) on the python code

I installed and formatted the code using black, thanks for the advice.

File: terraform/modules/pre-auth-lambda/iam.tf (hunk 2 above)
Comment (fernando-schmidt-dmw) [reply terraform/modules/pre-auth-lambda/iam.tf]: This is done.

**Comment by fernando-schmidt-dmw** [comment]
> You are missing the log group resource for the Lambda's logs

I created a new file called cloudwatch.tf inside which I declared this resource, can you please review it? Thanks.

File: terraform/modules/pre-auth-lambda/iam.tf [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+
+data aws_iam_policy_document policy_dynamodb_read_write_user_expiry {
+  statement {
+    actions = [
+        "dynamodb:PutItem",
+        "dynamodb:GetItem",
+        "dynamodb:Query",
+        "dynamodb:UpdateItem"
+    ]
+    resources = ["${aws_dynamodb_table.dynamodb_table_user.arn}"]
```
Comment (dan-hill2802) [inline terraform/modules/pre-auth-lambda/iam.tf:32]:
```suggestion
    resources = [aws_dynamodb_table.dynamodb_table_user.arn]
```

File: terraform/modules/pre-auth-lambda/lambda.tf [hunk 11]
```diff
@@ -0,0 +1,13 @@
+resource "aws_lambda_function" "lambda_user_expiry" {
```
[the remark this answers is not in the document]
Comment (fernando-schmidt-dmw) [reply terraform/modules/pre-auth-lambda/lambda.tf]: This is done.

File: terraform/modules/pre-auth-lambda/dynamodb.tf [hunk 12]
```diff
@@ -0,0 +1,10 @@
+resource "aws_dynamodb_table" "dynamodb_table_user" {
```
[the remark this answers is not in the document]
Comment (fernando-schmidt-dmw) [reply terraform/modules/pre-auth-lambda/dynamodb.tf:1]: This is done.

File: terraform/modules/pre-auth-lambda/iam.tf [hunk 13]
```diff
@@ -0,0 +1,50 @@
+resource aws_iam_role role_for_lambda_user_expiry {
```
[the remark this answers is not in the document]
Comment (fernando-schmidt-dmw) [reply terraform/modules/pre-auth-lambda/iam.tf:1]: This is done.

File: terraform/modules/pre-auth-lambda/iam.tf (hunk 10 above)
Comment (fernando-schmidt-dmw) [reply terraform/modules/pre-auth-lambda/iam.tf]: This is done.

File: terraform/modules/pre-auth-lambda/lambda/lambda_function.py [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+from dateutil.relativedelta import relativedelta
+import boto3
+from boto3.dynamodb.conditions import Key
+
+table_name = os.environ.get("TABLE_NAME")
+if table_name is None:
+    message = "Variable TABLE_NAME was not provided."
+    print(message)
+    raise Exception(message)
+region_name = os.environ.get("AWS_REGION")
```
Comment (liamraemclauchlan) [inline terraform/modules/pre-auth-lambda/lambda/lambda_function.py:12]: possibly a new line to aid in readability

File: terraform/modules/pre-auth-lambda/lambda.tf [hunk 15]
```diff
@@ -0,0 +1,14 @@
+resource "aws_lambda_function" "lambda_pre_auth" {
+  filename         = "${path.module}/${var.name_prefix}-pre-auth.zip"
```
Comment (dan-hill2802) [inline terraform/modules/pre-auth-lambda/lambda.tf:2]: this should reference `archive_file.lambda_pre_auth_zip.output_path`

File: terraform/modules/pre-auth-lambda/variables.tf [hunk 16]
```diff
@@ -0,0 +1,9 @@
...
+
+variable "common_tags" {
+  description = "common tags to apply to aws resources"
```
Comment (dan-hill2802) [inline terraform/modules/pre-auth-lambda/variables.tf:7]: you either need to add `(Required)` or add a default

File: terraform/modules/pre-auth-lambda/lambda/lambda_function.py (hunk 14 above)
Comment (fernando-schmidt-dmw) [reply terraform/modules/pre-auth-lambda/lambda/lambda_function.py:12]: This is done.

File: terraform/modules/pre-auth-lambda/variables.tf (hunk 16 above)
Comment (fernando-schmidt-dmw) [reply terraform/modules/pre-auth-lambda/variables.tf:7]: This is done.

File: terraform/modules/pre-auth-lambda/lambda.tf (hunk 15 above)
Comment (fernando-schmidt-dmw) [reply terraform/modules/pre-auth-lambda/lambda.tf]: This is done.

**Pull request merged**